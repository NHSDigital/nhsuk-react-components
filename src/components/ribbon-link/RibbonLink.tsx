/* eslint-disable @typescript-eslint/no-redeclare */
import React, { HTMLProps } from 'react';
import classNames from 'classnames';
import { ArrowRightCircleIcon } from 'nhsuk-react-components';
import Bar from './components/Bar';

type RibbonFlavours = 'hot' | 'mild' | 'cool';

interface RibbonProps extends HTMLProps<HTMLButtonElement> {
  flavour: RibbonFlavours;
  type?: 'button' | 'submit' | 'reset';
}

interface RibbonLink extends React.FC<RibbonProps> {
  Bar: React.FC<HTMLProps<HTMLDivElement>>;
}

const RibbonLink: RibbonLink = ({ children, flavour, className, ...rest }) => (
  <button
    className={classNames('nhsuk-ribbon-link', `nhsuk-ribbon-link--${flavour}`, className)}
    type="button"
    {...rest}
  >
    <ArrowRightCircleIcon className="nhsuk-ribbon-link__icon" />
    {children}
  </button>
);

RibbonLink.Bar = Bar;

export default RibbonLink;
