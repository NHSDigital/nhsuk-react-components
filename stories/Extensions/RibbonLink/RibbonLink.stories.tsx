/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import RibbonLink from '@components/ribbon-link/RibbonLink';

const meta: Meta<typeof RibbonLink> = {
  title: 'Extensions/ RibbonLink',
  component: RibbonLink,
};
export default meta;

type Story = StoryObj<typeof RibbonLink>;

export const CoolRibbon: Story = {
  render: () => (
    <div className="ribbonlink-demo">
      <RibbonLink flavour="cool">Cool Ribbon</RibbonLink>
    </div>
  ),
};

export const MildRibbon: Story = {
  render: () => (
    <div className="ribbonlink-demo">
      <RibbonLink flavour="mild">Mild Ribbon</RibbonLink>
    </div>
  ),
};

export const HotRibbon: Story = {
  render: () => (
    <div className="ribbonlink-demo">
      <RibbonLink flavour="hot">Hot Ribbon</RibbonLink>
    </div>
  ),
};

export const RibbonLinkBar: Story = {
  render: () => (
    <RibbonLink.Bar className="ribbonlink-demo">
      <RibbonLink flavour="cool">Cool Ribbon</RibbonLink>
      <RibbonLink flavour="mild">Mild Ribbon</RibbonLink>
      <RibbonLink flavour="hot">Hot Ribbon</RibbonLink>
    </RibbonLink.Bar>
  ),
};
