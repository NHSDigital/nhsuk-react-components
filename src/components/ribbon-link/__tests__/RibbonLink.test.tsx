import React from 'react';
import {render} from '@testing-library/react'
import RibbonLink from '..';

describe('RibbonLink', () => {
  it('matches snapshot', () => {
    const coolRibbon = render(<RibbonLink flavour="cool">Cool</RibbonLink>);
    const mildRibbon = render(<RibbonLink flavour="mild">Mild</RibbonLink>);
    const hotRibbon = render(<RibbonLink flavour="hot">Hot</RibbonLink>);

    expect(coolRibbon.container.querySelector('button')?.className).toBe('nhsuk-ribbon-link nhsuk-ribbon-link--cool');
    expect(mildRibbon.container.querySelector('button')?.className).toBe('nhsuk-ribbon-link nhsuk-ribbon-link--mild');
    expect(hotRibbon.container.querySelector('button')?.className).toBe('nhsuk-ribbon-link nhsuk-ribbon-link--hot');

    expect(coolRibbon).toMatchSnapshot();
    expect(mildRibbon).toMatchSnapshot();
    expect(hotRibbon).toMatchSnapshot();
  });

  describe('RibbonLink.Bar', () => {
    it('matches snapshot', () => {
      const bar = render(<RibbonLink.Bar />);
      expect(bar).toMatchSnapshot();
    });
  });
});
