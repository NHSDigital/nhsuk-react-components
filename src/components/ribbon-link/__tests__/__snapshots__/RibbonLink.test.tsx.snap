// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RibbonLink RibbonLink.Bar matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nhsuk-ribbon-link__bar"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="nhsuk-ribbon-link__bar"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RibbonLink matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--cool"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Cool
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--mild"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Mild
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--hot"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Hot
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="nhsuk-ribbon-link nhsuk-ribbon-link--cool"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
        height="34"
        viewBox="0 0 24 24"
        width="34"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
        />
      </svg>
      Cool
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RibbonLink matches snapshot 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--cool"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Cool
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--mild"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Mild
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--hot"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Hot
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="nhsuk-ribbon-link nhsuk-ribbon-link--mild"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
        height="34"
        viewBox="0 0 24 24"
        width="34"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
        />
      </svg>
      Mild
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RibbonLink matches snapshot 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--cool"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Cool
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--mild"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Mild
      </button>
    </div>
    <div>
      <button
        class="nhsuk-ribbon-link nhsuk-ribbon-link--hot"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
          height="34"
          viewBox="0 0 24 24"
          width="34"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
          />
        </svg>
        Hot
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="nhsuk-ribbon-link nhsuk-ribbon-link--hot"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__arrow-right-circle nhsuk-ribbon-link__icon"
        height="34"
        viewBox="0 0 24 24"
        width="34"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z"
        />
      </svg>
      Hot
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
